File: 2.0/watchOS/WatchSprings-2.0/WatchSprings Watch App/screens/BattlePassTrailerScreen.swift
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.


import SwiftUI
import AVKit
import AVFoundation
import Combine

struct BattlePassTrailerScreen: View
{
    @State var viewModel: BattlePassViewModel
    
    @Environment(\.dismiss) private var dismiss
   
    @State private var player: AVPlayer? = nil
    @State private var playbackObserver: AnyCancellable?

    var body: some View
    {
        ChildScreen(title: "Trailer", isLoading: viewModel.isLoading)
        {
            VStack
            {
                if let urlString = viewModel.selectedTrailerUrl,
                   let videoURL = URL(string: urlString)
                {
                    VideoPlayer(player: player)
                        .onAppear {
                            let avPlayer = AVPlayer(url: videoURL)
                            player = avPlayer
                            avPlayer.play()
                            
                            playbackObserver = NotificationCenter.default
                                .publisher(for: .AVPlayerItemDidPlayToEndTime, object: avPlayer.currentItem)
                                .sink { _ in dismiss() }
                        }
                        .onDisappear {
                            playbackObserver?.cancel()
                            playbackObserver = nil
                        }
                        .frame(height: 120)
                    
                    Button("Close") {
                        dismiss()
                    }
                    .buttonStyle(.bordered)
                    
                } else {
                    Text("No trailer available")
                        .foregroundColor(.gray)
                }
            }
        }
    }
}

File: 2.0/WearOS/WatchSprings-2.0/app/src/main/java/copyright/markmorris/allrightsreserved/watchsprings/screens/BattlePassTrailerScreen.kt
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

package copyright.markmorris.allrightsreserved.watchsprings

import android.view.Gravity
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.annotation.OptIn
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.net.toUri
import androidx.media3.common.MediaItem
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.ui.AspectRatioFrameLayout
import androidx.media3.ui.PlayerView
import androidx.navigation.NavHostController
import androidx.wear.compose.material.Button
import androidx.wear.compose.material.MaterialTheme
import androidx.wear.compose.material.Text

@OptIn(UnstableApi::class)
@Composable
fun BattlePassTrailerScreen(navController: NavHostController, viewModel: BattlePassViewModel)
{
    val context = LocalContext.current
    val trailerUrl = viewModel.selectedTrailerUrl.collectAsState().value
    val isRound = LocalConfiguration.current.isScreenRound

    ChildScreen(
        navController = null,
        title = "Trailer"
    ) {
        if (trailerUrl.isNullOrBlank()) {
            Text("No trailer selected.")
            return@ChildScreen
        }

        val exoPlayer = remember(context) {
            ExoPlayer.Builder(context).build().apply {
                setMediaItem(MediaItem.fromUri(trailerUrl.toUri()))
                prepare()
                playWhenReady = true
            }
        }

        DisposableEffect(Unit) {
            onDispose {
                exoPlayer.release()
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(20.dp)
                .padding(bottom = 4.dp),
            horizontalArrangement = Arrangement.Center
        ) {
            Button(
                onClick = {
                    exoPlayer.stop()
                    navController.popBackStack()
                }
            ) {
                Text(
                    text = "Back",
                    style = MaterialTheme.typography.caption3,
                 //   fontSize = 8.sp,
                    modifier = Modifier.padding(bottom = 2.dp)
                )
            }
        }

        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(bottom = 24.dp)
        ) {
            AndroidView(
                factory = { ctx ->
                    PlayerView(ctx).apply {
                        player = exoPlayer
                        useController = true
                        resizeMode = AspectRatioFrameLayout.RESIZE_MODE_FIT
                        layoutParams = FrameLayout.LayoutParams(
                            ViewGroup.LayoutParams.MATCH_PARENT,
                            ViewGroup.LayoutParams.MATCH_PARENT
                        ).apply {
                            gravity = Gravity.CENTER_HORIZONTAL or Gravity.TOP
                        }
                    }
                },
                modifier = Modifier
                    .fillMaxSize()
                    .align(Alignment.Center)
            )
        }
    }
}
